File: backend/api/schemas/image_generation.py
from datetime import datetime

from pydantic import BaseModel, Field, model_validator

from backend.models.enums import (
    GenerationMode,
    ImageGenerationProvider,
    ImagePromptType,
    SeedStrategy,
)


class ImageGenerationToolPresetRequest(BaseModel):
    """创建或更新生图工具配置的请求体。"""

    name: str
    provider: ImageGenerationProvider = ImageGenerationProvider.COMFYUI
    prompt_type: ImagePromptType = ImagePromptType.NATURAL_LANGUAGE
    description: str | None = None
    is_default: bool = False
    capabilities: dict = Field(
        default_factory=lambda: {"features": ["txt2img"], "limits": {}}
    )
    bindings: dict = Field(
        default_factory=lambda: {"schema_version": 1, "bindings": []}
    )
    comfy_base_url: str | None = None
    workflow_json: str | None = None
    positive_node_id: str | None = None
    positive_input_name: str | None = "text"
    negative_node_id: str | None = None
    negative_input_name: str | None = None
    seed_node_id: str | None = None
    seed_input_name: str | None = None
    api_base_url: str | None = None
    endpoint_path: str | None = "/images/generations"
    api_key: str | None = None
    model: str | None = None
    size: str | None = "1024x1024"
    response_format: str | None = "b64_json"
    seed_field_name: str | None = None
    negative_prompt_field_name: str | None = None
    extra_body_json: str | None = None


class ImageGenerationToolPresetResponse(BaseModel):
    """生图工具配置响应体。"""

    id: int
    name: str
    provider: ImageGenerationProvider
    prompt_type: ImagePromptType
    description: str | None
    is_default: bool
    capabilities: dict
    bindings: dict
    comfy_base_url: str | None
    workflow_json: str | None
    positive_node_id: str | None
    positive_input_name: str | None
    negative_node_id: str | None
    negative_input_name: str | None
    seed_node_id: str | None
    seed_input_name: str | None
    api_base_url: str | None
    endpoint_path: str | None
    api_key: str | None
    model: str | None
    size: str | None
    response_format: str | None
    seed_field_name: str | None
    negative_prompt_field_name: str | None
    extra_body_json: str | None
    created_at: datetime
    updated_at: datetime


class ImageGenerationToolPresetListResponse(BaseModel):
    """生图工具配置列表响应体。"""

    items: list[ImageGenerationToolPresetResponse]


ComfyWorkflowPresetRequest = ImageGenerationToolPresetRequest
ComfyWorkflowPresetResponse = ImageGenerationToolPresetResponse
ComfyWorkflowPresetListResponse = ImageGenerationToolPresetListResponse


class ComicImageResponse(BaseModel):
    """生成图片响应体。"""

    id: int
    page_id: int
    generation_run_id: int | None
    image_url: str | None
    local_path: str | None
    seed: int | None
    workflow_name: str | None
    prompt: str | None
    negative_prompt: str | None
    score: float | None
    sha256: str | None
    width: int | None
    height: int | None
    is_selected: bool
    created_at: datetime


class ImageGenerationPageResponse(BaseModel):
    """图片生成页面列表中的单页状态。"""

    page_id: int
    page_no: int
    prompt_type: ImagePromptType | None
    positive_prompt: str | None
    status: str
    selected_image_id: int | None
    latest_spec_id: int | None = None
    spec_warnings: list[dict] = Field(default_factory=list)
    completed_candidates: int = 0
    images: list[ComicImageResponse]


class ImageGenerationPageListResponse(BaseModel):
    """脚本任务下可生成图片页面的响应。"""

    task_id: int
    project_id: int
    items: list[ImageGenerationPageResponse]


class GenerateImagesRequest(BaseModel):
    """批量或单页图片生成请求体。"""

    tool_preset_id: int | None = Field(default=None, gt=0)
    workflow_preset_id: int | None = Field(default=None, gt=0)
    poll_interval_seconds: float = Field(default=2.0, ge=0.5, le=20)
    wait_timeout_seconds: float = Field(default=600.0, ge=30, le=3600)
    candidates_per_page: int = Field(default=1, ge=1, le=4)
    generation_mode: GenerationMode = GenerationMode.PREVIEW
    seed_strategy: SeedStrategy = SeedStrategy.PER_PAGE

    @model_validator(mode="after")
    def validate_preset_id(self):
        """新字段 tool_preset_id 优先，兼容旧 workflow_preset_id。"""

        if self.tool_preset_id is None and self.workflow_preset_id is None:
            raise ValueError("tool_preset_id is required")
        return self

    @property
    def effective_tool_preset_id(self) -> int:
        """读取实际使用的工具配置 id。"""

        return self.tool_preset_id or self.workflow_preset_id or 0


class GenerationTaskResponse(BaseModel):
    """ComfyUI 生成任务响应体。"""

    id: int
    project_id: int
    page_id: int | None
    comfy_prompt_id: str | None
    status: str
    batch_size: int
    error_message: str | None
    created_at: datetime
    updated_at: datetime


class GenerationRunResponse(BaseModel):
    """单候选生成溯源响应；JSON 快照保持原始结构。"""

    id: int
    generation_task_id: int
    page_id: int
    image_spec_id: int
    tool_preset_id: int
    provider: ImageGenerationProvider
    prompt_type: ImagePromptType
    candidate_index: int
    seed: int | None
    seed_applied: bool
    seed_strategy: str
    generation_mode: str
    status: str
    external_request_id: str | None
    workflow: dict | None
    workflow_hash: str | None
    bindings: dict
    resolved_assets: list
    degradations: list
    applied_spec: dict
    error_code: str | None
    error_message: str | None
    created_at: datetime
    updated_at: datetime
    finished_at: datetime | None

File: frontend/src/api/imageGeneration.ts
import type { ScriptTask } from './scripts'
import { ApiError, apiHeaders, normalizeBackendError, parseApiErrorResponse } from './errors'

export type ImageGenerationTool = {
  id: number
  name: string
  provider: 'comfyui' | 'openai_images_compatible'
  prompt_type: 'tag' | 'natural_language' | 'hybrid'
  description: string | null
  comfy_base_url: string | null
  workflow_json: string | null
  is_default: boolean
  capabilities: { features?: string[]; limits?: Record<string, number> }
  bindings: { schema_version?: number; bindings?: Array<{ source: string; node_id: string; input_name: string }> }
  positive_node_id: string | null
  positive_input_name: string | null
  negative_node_id: string | null
  negative_input_name: string | null
  seed_node_id: string | null
  seed_input_name: string | null
  api_base_url: string | null
  endpoint_path: string | null
  api_key: string | null
  model: string | null
  size: string | null
  response_format: string | null
  seed_field_name: string | null
  negative_prompt_field_name: string | null
  extra_body_json: string | null
  created_at: string
  updated_at: string
}

export type ImageGenerationToolPayload = {
  name: string
  provider: 'comfyui' | 'openai_images_compatible'
  prompt_type: 'tag' | 'natural_language' | 'hybrid'
  description?: string | null
  is_default: boolean
  capabilities?: Record<string, unknown>
  bindings?: Record<string, unknown>
  comfy_base_url?: string | null
  workflow_json?: string | null
  positive_node_id?: string | null
  positive_input_name?: string | null
  negative_node_id?: string | null
  negative_input_name?: string | null
  seed_node_id?: string | null
  seed_input_name?: string | null
  api_base_url?: string | null
  endpoint_path?: string | null
  api_key?: string | null
  model?: string | null
  size?: string | null
  response_format?: string | null
  seed_field_name?: string | null
  negative_prompt_field_name?: string | null
  extra_body_json?: string | null
}

export type GeneratedImage = {
  id: number
  page_id: number
  generation_run_id: number | null
  image_url: string | null
  local_path: string | null
  seed: number | null
  workflow_name: string | null
  prompt: string | null
  negative_prompt: string | null
  score: number | null
  sha256: string | null
  width: number | null
  height: number | null
  is_selected: boolean
  created_at: string
}

export type ImageGenerationPage = {
  page_id: number
  page_no: number
  prompt_type: 'tag' | 'natural_language' | 'hybrid' | null
  positive_prompt: string | null
  status: string
  selected_image_id: number | null
  latest_spec_id: number | null
  spec_warnings: Array<{ code?: string; message?: string }>
  completed_candidates: number
  images: GeneratedImage[]
}

export type GenerateImagesPayload = {
  tool_preset_id: number
  poll_interval_seconds: number
  wait_timeout_seconds: number
  candidates_per_page: number
  generation_mode: 'preview' | 'final'
  seed_strategy: 'per_page' | 'shared_candidate'
}

export type GenerationRun = {
  id: number
  generation_task_id: number
  page_id: number
  image_spec_id: number
  tool_preset_id: number
  provider: 'comfyui' | 'openai_images_compatible'
  prompt_type: 'tag' | 'natural_language' | 'hybrid'
  candidate_index: number
  seed: number | null
  seed_applied: boolean
  seed_strategy: string
  generation_mode: string
  status: string
  external_request_id: string | null
  workflow: Record<string, unknown> | null
  workflow_hash: string | null
  bindings: Record<string, unknown>
  resolved_assets: unknown[]
  degradations: unknown[]
  applied_spec: Record<string, unknown>
  error_code: string | null
  error_message: string | null
  created_at: string
  updated_at: string
  finished_at: string | null
}

export type GenerationTask = {
  id: number
  project_id: number
  page_id: number | null
  comfy_prompt_id: string | null
  status: string
  batch_size: number
  error_message: string | null
  created_at: string
  updated_at: string
}

type SseEvent = {
  event: string
  data: string
}

export type ImageGenerationStreamCallbacks = {
  onEvent: (event: string, payload: Record<string, unknown>) => void
  onError: (error: ApiError) => void
}

const requestJson = async <T>(url: string, options: RequestInit = {}): Promise<T> => {
  const response = await fetch(url, {
    ...options,
    headers: apiHeaders(options.headers),
  })

  if (!response.ok) {
    throw await parseApiErrorResponse(response)
  }

  if (response.status === 204) {
    return undefined as T
  }

  return (await response.json()) as T
}

export const listImageGenerationTools = async (): Promise<ImageGenerationTool[]> => {
  const result = await requestJson<{ items: ImageGenerationTool[] }>('/api/image-generation/tools')
  return result.items
}

export const createImageGenerationTool = (
  payload: ImageGenerationToolPayload,
): Promise<ImageGenerationTool> =>
  requestJson<ImageGenerationTool>('/api/image-generation/tools', {
    method: 'POST',
    body: JSON.stringify(payload),
  })

export const updateImageGenerationTool = (
  toolId: number,
  payload: ImageGenerationToolPayload,
): Promise<ImageGenerationTool> =>
  requestJson<ImageGenerationTool>(`/api/image-generation/tools/${toolId}`, {
    method: 'PUT',
    body: JSON.stringify(payload),
  })

export const deleteImageGenerationTool = (toolId: number): Promise<void> =>
  requestJson<void>(`/api/image-generation/tools/${toolId}`, {
    method: 'DELETE',
  })

export const listImageGenerationPages = async (
  taskId: number,
  options: {
    promptType?: 'tag' | 'natural_language' | 'hybrid'
    generationMode?: 'preview' | 'final'
  } = {},
): Promise<ImageGenerationPage[]> => {
  const params = new URLSearchParams()
  if (options.promptType) params.set('prompt_type', options.promptType)
  if (options.generationMode) params.set('generation_mode', options.generationMode)
  const query = params.size ? `?${params.toString()}` : ''
  const result = await requestJson<{ items: ImageGenerationPage[] }>(
    `/api/image-generation/script-tasks/${taskId}/pages${query}`,
  )
  return result.items
}

export const suspendImageGenerationTask = (taskId: number): Promise<GenerationTask> =>
  requestJson<GenerationTask>(`/api/image-generation/tasks/${taskId}/suspend`, {
    method: 'POST',
  })

export const selectGeneratedImage = (
  pageId: number,
  imageId: number,
): Promise<ImageGenerationPage> =>
  requestJson<ImageGenerationPage>(`/api/image-generation/pages/${pageId}/images/${imageId}/select`, {
    method: 'POST',
  })

export const getGenerationRun = (runId: number): Promise<GenerationRun> =>
  requestJson<GenerationRun>(`/api/image-generation/runs/${runId}`)

export const streamGenerateImagesForTask = (
  taskId: number,
  payload: GenerateImagesPayload,
  callbacks: ImageGenerationStreamCallbacks,
): Promise<void> =>
  streamSse(`/api/image-generation/script-tasks/${taskId}/generate/stream`, payload, callbacks)

export const streamContinueImagesForTask = (
  taskId: number,
  payload: GenerateImagesPayload,
  callbacks: ImageGenerationStreamCallbacks,
): Promise<void> =>
  streamSse(`/api/image-generation/script-tasks/${taskId}/continue/stream`, payload, callbacks)

export const streamGenerateImagesForPage = (
  pageId: number,
  payload: GenerateImagesPayload,
  callbacks: ImageGenerationStreamCallbacks,
): Promise<void> =>
  streamSse(`/api/image-generation/pages/${pageId}/generate/stream`, payload, callbacks)

const streamSse = async (
  url: string,
  payload: GenerateImagesPayload,
  callbacks: ImageGenerationStreamCallbacks,
): Promise<void> => {
  const response = await fetch(url, {
    method: 'POST',
    headers: apiHeaders(),
    body: JSON.stringify(payload),
  })

  if (!response.ok || response.body === null) {
    throw await parseApiErrorResponse(response)
  }

  const reader = response.body.getReader()
  const decoder = new TextDecoder()
  let buffer = ''

  const handleEvent = (event: SseEvent) => {
    const parsedPayload = event.data ? JSON.parse(event.data) : {}
    if (event.event === 'error') {
      const backendError = normalizeBackendError(parsedPayload)
      callbacks.onError(
        new ApiError(backendError.message || 'Image generation stream error', {
          code: backendError.code,
          payload: parsedPayload,
        }),
      )
      return
    }
    callbacks.onEvent(event.event, parsedPayload as Record<string, unknown>)
  }

  while (true) {
    const { done, value } = await reader.read()
    if (done) {
      break
    }

    buffer += decoder.decode(value, { stream: true })
    const chunks = buffer.split(/\r?\n\r?\n/)
    buffer = chunks.pop() ?? ''

    for (const chunk of chunks) {
      const event = parseSseChunk(chunk)
      if (event !== null) {
        handleEvent(event)
      }
    }
  }

  const tail = parseSseChunk(buffer)
  if (tail !== null) {
    handleEvent(tail)
  }
}

const parseSseChunk = (chunk: string): SseEvent | null => {
  const lines = chunk
    .split('\n')
    .map((line) => line.trimEnd())
    .filter((line) => line.length > 0)

  if (lines.length === 0) {
    return null
  }

  let event = 'message'
  const dataLines: string[] = []

  for (const line of lines) {
    if (line.startsWith(':')) {
      continue
    }
    if (line.startsWith('event:')) {
      event = line.slice('event:'.length).trim()
    } else if (line.startsWith('data:')) {
      dataLines.push(line.slice('data:'.length).trimStart())
    }
  }

  if (dataLines.length === 0) {
    return null
  }

  return {
    event,
    data: dataLines.join('\n'),
  }
}

export type { ScriptTask }
